test/test_utils.jl

```diff
@@ -0,0 +1,394 @@
+    @test isapprox(jacb[1], jac_exact[2], rtol=1e-10)
+    @test xvec == xvec2
+    @test yvec == yvec2
+end
+
+function test_hessians(backend; multiple_inputs=false)
+    if multiple_inputs
+        # ... but 
+        error("multiple_inputs=true is not supported.")
+    else
+        # explicit test that AbstractDifferentiation throws an error
+        # don't support tuple of Hessians
+        @test_throws AssertionError H1 = AD.hessian(backend, fgrad, (xvec, yvec))
+        @test_throws MethodError H1 = AD.hessian(backend, fgrad, xvec, yvec)
+    end
+   
+    # @test dfgraddxdx(xvec,yvec) ≈ H1[1] atol=1e-10
+    # @test dfgraddydy(xvec,yvec) ≈ H1[2] atol=1e-10
+
+    # test if single input (no tuple works)
+    fhess = x -> fgrad(x, yvec)
+    hess1 = AD.hessian(backend, fhess, xvec)
+    # test with respect to analytical solution
+    @test dfgraddxdx(xvec,yvec) ≈ hess1[1] atol=1e-10
+    
+    valscalar, hess2 = AD.value_and_hessian(backend, fhess, xvec)
+    @test valscalar == fgrad(xvec, yvec)
+    @test norm.(hess2 .- hess1) == (0,)
+    valscalar, grad, hess3 = AD.value_gradient_and_hessian(backend, fhess, xvec)
+    @test valscalar == fgrad(xvec, yvec)
+    @test norm.(grad .- AD.gradient(backend, fhess, xvec)) == (0,)
+    @test norm.(hess3 .- hess1) == (0,)
+    
+    @test xvec == xvec2
+    @test yvec == yvec2
+    fhess2 = x-> dfgraddx(x, yvec)
+    hess4 = AD.jacobian(backend, fhess2, xvec)
+    @test minimum(isapprox.(hess4, hess1, atol=1e-10))
+end
+
+function test_jvp(backend; multiple_inputs=true, vaugmented=false)
+    v = (rand(length(xvec)), rand(length(yvec)))
+
+    if multiple_inputs
+        if vaugmented
+            identity_like = AD.identity_matrix_like(v)
+            vaug = map(identity_like) do identity_like_i
+                identity_like_i .* v
+            end
+
+            pf1 = map(v->AD.pushforward_function(backend, fjac, xvec, yvec)(v), vaug)
+            ((valvec1, pf2x), (valvec2, pf2y)) = map(v->AD.value_and_pushforward_function(backend, fjac, xvec, yvec)(v), vaug)
+        else
+            pf1 = AD.pushforward_function(backend, fjac, xvec, yvec)(v)
+            valvec, pf2 = AD.value_and_pushforward_function(backend, fjac, xvec, yvec)(v) 
+            ((valvec1, pf2x), (valvec2, pf2y)) = (valvec, pf2[1]), (valvec, pf2[2])
+        end
+       
+        @test valvec1 == fjac(xvec, yvec)
+        @test valvec2 == fjac(xvec, yvec)
+        @test norm.((pf2x,pf2y) .- pf1) == (0, 0)
+        # test with respect to analytical solution
+        @test minimum(isapprox.(pf1, (jxvp(xvec,yvec,v[1]), jyvp(xvec,yvec,v[2])), atol=1e-10))
+        @test xvec == xvec2
+        @test yvec == yvec2
+    end
+
+    valvec1, pf1 = AD.value_and_pushforward_function(backend, x -> fjac(x, yvec), xvec)(v[1])
+    valvec2, pf2 = AD.value_and_pushforward_function(backend, y -> fjac(xvec, y), yvec)(v[2])
+
+    # NOTE: added for compatibility with FDMBackend2. Why is this needed?
```

Comment:
@mohamed82008 do you know this special-casing is needed?

Reply:
it shouldn't be needed

Reply:
Seems to be though (see latest commit and failing tests).